File: tests/csharp/csharp_classes/class_method_chaining.cs
// vybe-test: csharp/csharp_classes/class_method_chaining
// origin: languages/csharp/tests/csharp/test_csharp_classes.rs

using static __Harness;

var r = new Builder().Add("a").Add("b").Add("c").Build();
__P((r).ToString());
__Check("abc");

class Builder {
    private string result = "";
    public Builder Add(string s) {
        result += s;
        return this;
    }
    public string Build() { return result; }
}

public static class __Harness {
    public static string __buf = "";
    public static void __P(string s) { __buf = __buf + s + "\n"; }
    public static void __Pr(string s) { __buf = __buf + s; }
    public static void __Check(string want) {
        if (__buf != want && __buf != want + "\n") {
            Console.WriteLine("FAIL: want [" + want + "] got [" + __buf + "]");
            throw new Exception("assertion failed");
        }
    }
}

File: tests/js/ecma_classes/class_method_chaining.js
// vybe-test: js/ecma_classes/class_method_chaining
// origin: languages/js/tests/js/test_ecma_classes.rs

function __line(...args) {
    // console.log joins its arguments with a single space. String() is the
    // coercion Vybe's logging host applies to each one.
    return args.map(String).join(" ");
}

function __check(got, want) {
    if (got !== want) {
        console.log("FAIL: want [" + want + "] got [" + got + "]");
        throw new Error("assertion failed");
    }
}

class Builder {
    constructor() { this.parts = []; }
    add(part) { this.parts.push(part); return this; }
    build() { return this.parts.join(", "); }
}
const result = new Builder().add("a").add("b").add("c").build();
__check(__line(result), "a, b, c");
